instrumentation/vaadin-14.2/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/vaadin/VaadinTracer.java

```diff
@@ -0,0 +1,163 @@
+import io.opentelemetry.context.Context;
+import io.opentelemetry.context.ContextKey;
+import io.opentelemetry.instrumentation.api.servlet.ServletContextPath;
+import io.opentelemetry.instrumentation.api.tracer.BaseTracer;
+import io.opentelemetry.instrumentation.api.tracer.ServerSpan;
+import java.lang.reflect.Method;
+
+public class VaadinTracer extends BaseTracer {
+  private static final ContextKey<VaadinServiceContext> SERVICE_CONTEXT_KEY =
+      ContextKey.named("opentelemetry-vaadin-service-context-key");
+  private static final ContextKey<Object> REQUEST_HANDLER_CONTEXT_KEY =
+      ContextKey.named("opentelemetry-vaadin-request-handler-context-key");
+
+  private static final VaadinTracer TRACER = new VaadinTracer();
+
+  public static VaadinTracer tracer() {
+    return TRACER;
+  }
+
+  private VaadinTracer() {
+    super(GlobalOpenTelemetry.get());
+  }
+
+  public Context startVaadinServiceSpan(VaadinService vaadinService, Method method) {
+    String spanName = spanNameForMethod(vaadinService.getClass(), method);
+    Context context = super.startSpan(spanName);
+    return context.with(SERVICE_CONTEXT_KEY, new VaadinServiceContext(spanName));
+  }
+
+  public void endSpan(Context context, Throwable throwable) {
+    if (throwable != null) {
+      endExceptionally(context, throwable);
+    } else {
+      end(context);
+    }
+  }
+
+  public void endVaadinServiceSpan(Context context, Throwable throwable) {
+    endSpan(context, throwable);
+
+    VaadinServiceContext vaadinServiceContext = context.get(SERVICE_CONTEXT_KEY);
+    if (!vaadinServiceContext.isRequestHandled()) {
+      // none of the request handlers processed the request
+      // as we update server span name on call to each request handler currently server span name
+      // is set based on the last request handler even when it didn't process the request, set
+      // server span name to main request processing method name
+      Span span = ServerSpan.fromContextOrNull(context);
+      if (span != null) {
+        span.updateName(vaadinServiceContext.vaadinServiceSpanName);
+      }
+    }
+  }
+
+  public Context startRequestHandlerSpan(RequestHandler requestHandler, Method method) {
+    Context current = Context.current();
+    // ignore nested request handlers
+    if (current.get(REQUEST_HANDLER_CONTEXT_KEY) != null) {
+      return null;
+    }
+
+    String spanName = spanNameForMethod(requestHandler.getClass(), method);
+    VaadinServiceContext vaadinServiceContext = current.get(SERVICE_CONTEXT_KEY);
+    if (vaadinServiceContext != null && !vaadinServiceContext.isRequestHandled()) {
+      Span span = ServerSpan.fromContextOrNull(current);
+      if (span != null) {
+        // set server span name to request handler name
+        // we don't really know whether this request handler is going to be the one
+        // that process the request, if it isn't then next handler will also update
+        // server span name
```

Comment:
Do handlers call each other (and thus they are nested) or are they called one after another?

Reply:
As far as I understand both can happen. Main request processing calls handlers sequentially, but there are some handlers that call other handlers. Our instrumentation should only care about handlers in main processing and ignore the rest.